### PR TITLE
Add `config.generators.after_generate` for processing to generated files

### PR DESCRIPTION
**Pull request by y-yagi**
This is useful if users want to process generated files.
For example, can execute an autocorrect of RuboCop for generated files as like following.

```ruby
config.generators.after_generate do |files|
  system("bundle exec rubocop --auto-correct " + files.join(" "), exception: true)
end
```

### CODE REVIEW

**Comment by simi** [comment]
This covers a little difference use-case than I have seen benefit of in original #32996. I was looking for way to open (for example) migration file every-time I create it using generator, but do not enforce the same for all developers on same project. Also I would welcome to set this once in my "env" and have this enabled for all Rails apps having this capability (compatible version).

If I understand it well currently there's no easy way to achieve this, since (AFAIK) the global file `.railsrc` works only for `rails new` command. The only way I can think of for now is to have custom initializer globally git ignored and ensure it is copied in every Rails app I work on.

Thinking of my requirement, I think it should be possible to modify behaviour of rails generator commands when used in "non-interactive"(`!$stdout.tty?`) way and instead of user friendly output just provide list of generated files. Than it would be possible to do `bin/rails g migration my_tiny_winie_migration | xargs gvim`. WDYT?

**Comment by y-yagi** [comment]
Thanks for sharing your use-case!

It seems the requirement is different from my requirement.
My requirement is enforcing the style to the same for all developers on the same project.
So I think it's hard to fix all use-case in one fix. I keep this PR as is (to fix my use-case).

> bin/rails g migration my_tiny_winie_migration | xargs gvim 

This use-case can support by this PR. For example, only run `after_generate` when some env key. 

```ruby 
config.generators.after_generate do |files|
  if ENV["EDITOR_FOR_GENERATOR"]
    system(ENV["EDITOR_FOR_GENERATOR"], *files, exception: true)
  end
end 
``` 

```bash 
$ EDITOR_FOR_GENERATOR=vim ./bin/rails g model user name:string 
```  

Of course, this is a little ugly. That support is out of scope this PR,  but being able to pass a list of files to stdout in some way is worth considering in other PR I think.

**Comment by simi** [comment]
>  only run after_generate when some env key

Yeah, but that should be added to every app just because of me.

I'll think about my use-case a little more and open separate PR if needed. Code changes in here would be beneficial for that as well.

**Comment by kaspth** [comment]
Not too fond of the stated use case as justification for adding it. Wouldn't you have to run Rubocop before pushing the files (e.g. after someone's edited it) or run a Rubocop check on CI anyway? But the possibilities that `after_generate` has are interesting enough, so let's go for it.

I think the `--edit` (which then uses `ENV["EDITOR"]`) would be great to have built-in, I've wanted that in the past too.

File: railties/lib/rails/generators/actions/create_migration.rb [hunk 1]
```diff
@@ -19,6 +19,11 @@ def identical?
           exists? && File.binread(existing_migration) == render
         end
 
+        def invoke!
+          invoked_file = super
+          File.exist?(@destination) ? invoked_file : relative_existing_migration
```
Comment (kaspth) [inline railties/lib/rails/generators/actions/create_migration.rb:24]: Why is this needed?

File: railties/lib/rails/generators.rb [hunk 2]
```diff
@@ -269,6 +275,7 @@ def invoke(namespace, args = ARGV, config = {})
         if klass = find_by_namespace(names.pop, names.any? && names.join(":"))
           args << "--help" if args.empty? && klass.arguments.any?(&:required?)
           klass.start(args, config)
+          run_after_generate_callback if config[:behavior] == :invoke
```
Comment (kaspth) [inline railties/lib/rails/generators.rb:278]: Does `start` return a value that we need to pass on? E.g. should we do `start(args, config).tap do`?

File: railties/lib/rails/generators.rb [hunk 3]
```diff
@@ -29,6 +29,7 @@ module Generators
     autoload :TestCase,        "rails/generators/test_case"
 
     mattr_accessor :namespace
+    mattr_reader :generated_files, default: []
```
Comment (kaspth) [inline railties/lib/rails/generators.rb:32]: Hm, not liking exposing this here. Can we manage everything entirely within `add_generated_file`? Maybe via `(@@generated_files ||= []) << file` or something?

If not, I'd move the `_reader` above the `_accessor` since that's the style we follow, iirc.

File: railties/lib/rails/generators/actions/create_migration.rb (hunk 1 above)
Comment (y-yagi) [reply railties/lib/rails/generators/actions/create_migration.rb:24]: `CreateMigration` keeps an existing file when content is the same with an invoked file. In that case, we need to return an existing file name. If do not, `after_generate` callback will return a non-existing file name(invoked file name).

File: railties/lib/rails/generators.rb (hunk 2 above)
Comment (y-yagi) [reply railties/lib/rails/generators.rb:277]: The `start` returns all result of `all_command`.  `all_command` includes methods that do not relate with generating files(e.g. `check_class_collision`, `hook_for`). It is a little difficult to extract that only result of generating files.

File: railties/lib/rails/generators.rb (hunk 3 above)
Comment (y-yagi) [reply railties/lib/rails/generators.rb]: I fixed that manage everything entirely within `add_generated_file`.

File: railties/lib/rails/generators.rb [hunk 4]
```diff
@@ -281,6 +287,11 @@ def invoke(namespace, args = ARGV, config = {})
         end
       end
 
+      def add_generated_file(file) # :nodoc:
+        (@@generated_files ||= []) << file
```
Comment (jonathanhefner) [inline railties/lib/rails/generators.rb:291]: It feels a little strange to have this as global state.  I'm not sure how it will play with the [`:inline` option](https://github.com/rails/rails/blob/b69ac962adeef897d4d1ba9aac40b9501ce25659/railties/lib/rails/generators/actions.rb#L252-L259) added in #37516.  If there is no way for this to be instance state on the instantiated generator, what do you think about clearing `@@generated_files` after all callbacks are run?

Comment (y-yagi) [reply railties/lib/rails/generators.rb:291]: Thanks, your advice is correct and clearing `@@generated_files` was needed.  I fixed.

**Comment by kaspth** [comment]
Rad! ❤️💪
